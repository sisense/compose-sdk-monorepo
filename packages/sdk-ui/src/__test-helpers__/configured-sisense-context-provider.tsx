import { mockToken, mockUrl } from '@/__mocks__/msw';
import { SisenseContextProviderProps } from '@/props';
import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';

const contextProviderProps: SisenseContextProviderProps = {
  url: mockUrl,
  token: mockToken,
  defaultDataSource: 'Sample ECommerce',
  appConfig: {
    queryCacheConfig: { enabled: false },
    trackingConfig: { enabled: false },
  },
};

export const ConfiguredSisenseContextProvider: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  return <SisenseContextProvider {...contextProviderProps}>{children}</SisenseContextProvider>;
};
