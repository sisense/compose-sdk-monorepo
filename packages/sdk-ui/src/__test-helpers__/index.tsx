/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
