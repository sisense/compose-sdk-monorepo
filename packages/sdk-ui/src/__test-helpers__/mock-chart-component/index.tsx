// Mocks the Chart component and provides a way to access the rendered charts.
// Use this to simulate user interactions with the chart while actual Chart component is mocked.

import { ChartProps } from '@/props';
import { RenderedChartModel } from './rendered-chart-model';

export type ChartMocksManager = {
  /**
   * Replace the actual Chart component with a mocked version.
   */
  mockChartComponent: () => void;
  /**
   * Restore the actual Chart component.
   */
  unmockChartComponent: () => void;
  /**
   * List of models of rendered charts.
   * Use these models to simulate user interactions with the rendered charts.
   */
  renderedCharts: RenderedChartModel[];
  /**
   * Clear all rendered charts from the mocks.
   */
  clearMocks: () => void;
};

const renderedCharts: RenderedChartModel[] = [];

const mockingConfig = {
  shouldChartBeMocked: false,
};

vi.mock('@/chart', async () => {
  const actualModule = await vi.importActual<typeof import('@/chart')>('@/chart');

  return {
    ...actualModule,
    Chart: (props: ChartProps) => {
      if (!mockingConfig.shouldChartBeMocked) {
        return actualModule.Chart(props); // Use the actual Chart component
      }
      const chartModel = new RenderedChartModel(
        props,
        `Chart_${renderedCharts.length} (${props.chartType})`,
      );
      renderedCharts.push(chartModel);
      return chartModel.render();
    },
  };
});

const mockChartComponent = () => {
  clearMocks();
  mockingConfig.shouldChartBeMocked = true;
};

const unmockChartComponent = () => {
  mockingConfig.shouldChartBeMocked = false;
};

const clearMocks = () => {
  renderedCharts.splice(0, renderedCharts.length);
};

export const chartMocksManager: ChartMocksManager = {
  mockChartComponent,
  unmockChartComponent,
  clearMocks,
  renderedCharts,
};
