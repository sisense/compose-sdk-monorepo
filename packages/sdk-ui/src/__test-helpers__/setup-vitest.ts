import matchers, { TestingLibraryMatchers } from '@testing-library/jest-dom/matchers';
import { cleanup } from '@testing-library/react';
import 'vitest-canvas-mock';

import createFetchMock from 'vitest-fetch-mock';
import { vi } from 'vitest';

// Add types for matchers from @testing-library/jest-dom
// https://github.com/testing-library/jest-dom/issues/439#issuecomment-1536524120
declare module 'vitest' {
  interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Manually extend Vitest's expect() with methods from @testing-library/jest-dom
expect.extend(matchers);

// Clean up mounted components after each test
// https://testing-library.com/docs/react-testing-library/api/#cleanup
afterEach(() => {
  cleanup();
});

vi.stubGlobal('__PACKAGE_VERSION__', '0.0.0');

// Mock ResizeObserver
const ResizeObserverMock = vi.fn(() => ({
  observe: vi.fn(),
  unobserve: vi.fn(),
  disconnect: vi.fn(),
}));

vi.stubGlobal('ResizeObserver', ResizeObserverMock);

// enable fetch mocking but don't mock anything by default
// and let tests decide to mock or not
const fetchMocker = createFetchMock(vi);
fetchMocker.enableMocks();
fetchMocker.dontMock();
