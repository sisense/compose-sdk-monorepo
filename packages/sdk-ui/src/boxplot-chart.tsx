import { BoxplotChartProps } from './props';
import { Chart, shouldSkipSisenseContextWaiting } from './chart';
import { asSisenseComponent } from './decorators/component-decorators/as-sisense-component';
/**
 * A React component representing data in a way that visually describes the distribution, variability,
 * and center of a data set along an axis.
 * See [Boxplot Chart](https://docs.sisense.com/main/SisenseLinux/box-and-whisker-plot.htm) for more information.
 *
 * @example
 * An example of using the component to visualize the `Sample ECommerce` data source:
 * ```tsx
 * <BoxplotChart
 *   dataSet={DM.DataSource}
 *   dataOptions={{
 *     category: [DM.Category.Category],
 *     value: [DM.Commerce.Cost],
 *     boxType: 'iqr',
 *     outliersEnabled: true,
 *   }}
 *   filters={[
 *     filterFactory.members(DM.Category.Category, ['Calculators', 'DVD Players', 'Routers']),
 *     filterFactory.members(DM.Commerce.AgeRange, ['19-24']),
 *   ]}
 *   styleOptions={{
 *     seriesLabels: {
 *       enabled: true,
 *     }
 *   }}
 * />
 * ```
 *
 * <img src="media://boxplot-chart-example-1.png" width="600px" />
 * @param props - Boxplot chart properties
 * @returns Boxplot Chart component
 * @beta
 */
export const BoxplotChart = asSisenseComponent({
  componentName: 'BoxplotChart',
  shouldSkipSisenseContextWaiting,
})((props: BoxplotChartProps) => {
  return <Chart {...props} chartType="boxplot" />;
});
