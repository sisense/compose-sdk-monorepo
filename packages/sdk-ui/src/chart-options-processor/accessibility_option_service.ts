import merge from 'ts-deepmerge';
import { HighchartsOptionsInternal } from './chart_options_service';

/*
 * Applies highcharts accessibility options to the chart.
 */
export const applyAccessibilityToChart = (
  chartOptions: HighchartsOptionsInternal,
): HighchartsOptionsInternal => {
  // TODO - Disable accessibility for now until we review the options
  return merge(chartOptions, { accessibility: { enabled: false } });
};
