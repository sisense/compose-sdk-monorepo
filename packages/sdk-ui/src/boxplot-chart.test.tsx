import { render } from '@testing-library/react';
import { HighchartsOptions, SeriesType } from './chart-options-processor/chart-options-service';
import type { Data } from '@ethings-os/sdk-data';
import { BoxplotChart } from './boxplot-chart';
import { withBlurredRows } from './__test-helpers__';

const dataSet: Data = {
  columns: [
    {
      name: 'Category',
      type: 'text',
    },
    {
      name: 'Box Min',
      type: 'number',
    },
    {
      name: 'Box Median',
      type: 'number',
    },
    {
      name: 'Box Max',
      type: 'number',
    },
    {
      name: 'Whisker Min',
      type: 'number',
    },
    {
      name: 'Whisker Max',
      type: 'number',
    },
    {
      name: 'Outlier Count',
      type: 'number',
    },
    {
      name: 'Cost (Outliers)',
      type: 'number',
    },
  ],
  rows: [
    [
      'Apple Mac Desktops',
      168.3239288330078,
      335.3749694824219,
      577.9123840332031,
      -446.05875396728516,
      1192.295066833496,
      2,
      '1232.0960693359375,1408.0069580078125',
    ],
    [
      'Apple Mac Laptops',
      153.461296081543,
      281.7227478027344,
      559.2744293212891,
      -455.2584037780761,
      1167.9941291809082,
      3,
      '1181.6976318359375,1200,1222.219970703125',
    ],
  ],
};

const [
  category,
  boxMinValue,
  boxMedianValue,
  boxMaxValue,
  whiskerMinValue,
  whiskerMaxValue,
  outlierCount,
  outlierValue,
] = dataSet.columns;

const costValue = {
  name: 'Cost',
  type: 'number',
  // having mocked jaql method in order to allow generating iternal formulas
  jaql: () => ({}),
};

describe('Boxplot Chart', () => {
  it('render boxplot with custom data options that allow to pass all values', async () => {
    const { findByLabelText } = render(
      <BoxplotChart
        dataSet={dataSet}
        dataOptions={{
          category: [category],
          value: [
            boxMinValue,
            boxMedianValue,
            boxMaxValue,
            whiskerMinValue,
            whiskerMaxValue,
            outlierCount,
          ],
          outliers: [outlierValue],
          valueTitle: 'Some data title',
        }}
        onBeforeRender={(options: HighchartsOptions) => {
          expect(options).toMatchSnapshot();
          return options;
        }}
      />,
    );

    expect(await findByLabelText('chart-root')).toBeInTheDocument();
  });

  it('render boxplot with single value in data options', async () => {
    const { findByLabelText } = render(
      <BoxplotChart
        dataSet={dataSet}
        dataOptions={{
          category: [category],
          value: [costValue],
          boxType: 'iqr',
          outliersEnabled: true,
        }}
        onBeforeRender={(options: HighchartsOptions) => {
          expect(options).toMatchSnapshot();
          return options;
        }}
      />,
    );

    expect(await findByLabelText('chart-root')).toBeInTheDocument();
  });

  it('render boxplot without outlier points', async () => {
    const { findByLabelText } = render(
      <BoxplotChart
        dataSet={dataSet}
        dataOptions={{
          category: [category],
          value: [costValue],
          boxType: 'iqr',
          outliersEnabled: false,
        }}
        onBeforeRender={(options: HighchartsOptions) => {
          expect(options.series?.length).toBe(1);
          return options;
        }}
      />,
    );

    expect(await findByLabelText('chart-root')).toBeInTheDocument();
  });

  it('render boxplot with only values (without category)', async () => {
    const { findByLabelText } = render(
      <BoxplotChart
        dataSet={dataSet}
        dataOptions={{
          category: [],
          value: [costValue],
          boxType: 'iqr',
          outliersEnabled: true,
        }}
        onBeforeRender={(options: HighchartsOptions) => {
          expect((options.series?.[0] as SeriesType).data.length).toBe(1);
          return options;
        }}
      />,
    );

    expect(await findByLabelText('chart-root')).toBeInTheDocument();
  });

  it('render boxplot with highlights', async () => {
    const bluredDataIndexes = [0];

    const { findByLabelText } = render(
      <BoxplotChart
        dataSet={withBlurredRows(dataSet, bluredDataIndexes)}
        dataOptions={{
          category: [category],
          value: [costValue],
          boxType: 'iqr',
          outliersEnabled: true,
        }}
        onBeforeRender={(options: HighchartsOptions) => {
          (options.series?.[0] as SeriesType).data.forEach(({ selected }, index) => {
            const isCorrectBlurApplied = bluredDataIndexes.includes(index) ? selected : !selected;
            expect(isCorrectBlurApplied).toBe(true);
          });

          return options;
        }}
      />,
    );

    expect(await findByLabelText('chart-root')).toBeInTheDocument();
  });
});
