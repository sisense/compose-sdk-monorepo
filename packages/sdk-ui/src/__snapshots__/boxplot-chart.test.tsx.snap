// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Boxplot Chart > render boxplot with custom data options that allow to pass all values 1`] = `
{
  "accessibility": {
    "enabled": false,
  },
  "boost": {
    "useGPUTranslations": true,
    "usePreAllocated": true,
  },
  "chart": {
    "animation": {
      "duration": 300,
    },
    "backgroundColor": "#FFFFFF",
    "events": {
      "load": [Function],
      "selection": [Function],
    },
    "polar": false,
    "spacing": [
      25,
      20,
      20,
      20,
    ],
    "type": "boxplot",
    "zooming": {
      "type": "x",
    },
  },
  "drilldown": {
    "activeDataLabelStyle": {
      "color": "#000000",
    },
    "breadcrumbs": {
      "buttonTheme": {
        "fill": "#ffcb05",
        "states": {
          "hover": {
            "fill": "#f2b900",
            "stroke": "#f2b900",
          },
        },
        "stroke": "#ffcb05",
        "style": {
          "color": "#3a4356",
        },
      },
    },
  },
  "legend": {
    "align": "center",
    "backgroundColor": "transparent",
    "borderColor": "transparent",
    "borderWidth": 0,
    "enabled": true,
    "itemStyle": {
      "color": "#000000",
      "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
      "fontSize": "13px",
      "fontWeight": "normal",
      "pointerEvents": "auto",
      "textOutline": "none",
    },
    "layout": "horizontal",
    "symbolRadius": 0,
    "verticalAlign": "bottom",
  },
  "plotOptions": {
    "boxplot": {
      "dataLabels": {
        "align": "center",
        "enabled": false,
        "formatter": [Function],
        "rotation": 0,
        "style": {
          "fontSize": "13px",
          "fontWeight": "normal",
        },
      },
    },
    "scatter": {
      "dataLabels": {
        "enabled": false,
      },
    },
    "series": {
      "animation": {
        "duration": 600,
      },
      "boostThreshold": 0,
      "dataLabels": {
        "align": "center",
        "enabled": false,
        "formatter": [Function],
        "rotation": 0,
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
        },
      },
      "grouping": false,
      "point": {
        "events": {
          "click": [Function],
        },
      },
      "softThreshold": false,
      "stickyTracking": false,
      "turboThreshold": 0,
    },
  },
  "series": [
    {
      "color": "#0090a1",
      "data": [
        {
          "blur": false,
          "color": "#0090a1",
          "custom": {
            "xValue": [
              "Apple Mac Desktops",
            ],
          },
          "fillColor": "#00cee6",
          "high": 1192.295066833496,
          "innerBoxColor": "#00afc4",
          "low": -446.05875396728516,
          "median": 335.3749694824219,
          "q1": 168.3239288330078,
          "q3": 577.9123840332031,
          "selected": false,
        },
        {
          "blur": false,
          "color": "#0090a1",
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "fillColor": "#00cee6",
          "high": 1167.9941291809082,
          "innerBoxColor": "#00afc4",
          "low": -455.2584037780761,
          "median": 281.7227478027344,
          "q1": 153.461296081543,
          "q3": 559.2744293212891,
          "selected": false,
        },
      ],
      "fillOpacity": 0.8,
      "legendIndex": 0,
      "maxPointWidth": 42,
      "medianWidth": 1,
      "minPointWidth": 20,
      "name": "Some data title",
      "showInLegend": true,
      "strokeOpacity": 0.8,
      "title": "Some data title",
      "whiskerLength": "100%",
      "whiskerWidth": 1,
      "yAxis": 0,
    },
    {
      "color": "#0090a1",
      "data": [
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Desktops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 0,
          "y": 1232.0960693359375,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Desktops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 0,
          "y": 1408.0069580078125,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 1,
          "y": 1181.6976318359375,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 1,
          "y": 1200,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 1,
          "y": 1222.219970703125,
        },
      ],
      "legendIndex": 1,
      "marker": {
        "enabled": true,
        "fillColor": "transparent",
        "lineColor": "#0090a1",
        "lineWidth": 1,
      },
      "name": "Some data title",
      "showInLegend": false,
      "title": "Some data title",
      "type": "scatter",
    },
  ],
  "title": {
    "text": null,
  },
  "tooltip": {
    "animation": false,
    "backgroundColor": "#FFFFFF",
    "borderColor": "#CCCCCC",
    "borderRadius": 10,
    "borderWidth": 1,
    "formatter": [Function],
    "style": {
      "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
    },
    "useHTML": true,
  },
  "xAxis": [
    {
      "categories": [
        "Apple Mac Desktops",
        "Apple Mac Laptops",
      ],
      "endOnTick": true,
      "gridLineColor": "#d1d1d7",
      "gridLineDashStyle": "Dot",
      "gridLineWidth": 0,
      "labels": {
        "autoRotation": [
          -10,
          -20,
          -30,
          -40,
          -50,
          -60,
          -70,
          -80,
          -90,
        ],
        "enabled": true,
        "formatter": [Function],
        "overflow": "none",
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
      },
      "lineColor": "#d1d1d7",
      "max": null,
      "min": null,
      "minorGridLineColor": "#d1d1d7",
      "minorGridLineDashStyle": "Dot",
      "minorTickColor": "#d1d1d7",
      "minorTickLength": 0,
      "plotBands": undefined,
      "startOnTick": true,
      "tickColor": "#d1d1d7",
      "tickLength": 0,
      "tickmarkPlacement": "on",
      "title": {
        "margin": 25,
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
        },
        "text": null,
      },
    },
  ],
  "yAxis": [
    {
      "gridLineColor": "#d1d1d7",
      "gridLineDashStyle": "Dot",
      "gridLineWidth": 1,
      "labels": {
        "autoRotation": [
          -10,
          -20,
          -30,
          -40,
          -50,
          -60,
          -70,
          -80,
          -90,
        ],
        "enabled": true,
        "formatter": [Function],
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
      },
      "lineColor": "#d1d1d7",
      "lineWidth": 1,
      "max": 1192.295066833496,
      "min": -455.2584037780761,
      "minorGridLineColor": "#d1d1d7",
      "minorGridLineDashStyle": "Dot",
      "minorGridLineWidth": 0,
      "minorTickColor": "#d1d1d7",
      "plotBands": undefined,
      "tickColor": "#d1d1d7",
      "tickInterval": null,
      "tickmarkPlacement": "on",
      "title": {
        "margin": 25,
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
        "text": null,
      },
    },
  ],
}
`;

exports[`Boxplot Chart > render boxplot with single value in data options 1`] = `
{
  "accessibility": {
    "enabled": false,
  },
  "boost": {
    "useGPUTranslations": true,
    "usePreAllocated": true,
  },
  "chart": {
    "animation": {
      "duration": 300,
    },
    "backgroundColor": "#FFFFFF",
    "events": {
      "load": [Function],
      "selection": [Function],
    },
    "polar": false,
    "spacing": [
      25,
      20,
      20,
      20,
    ],
    "type": "boxplot",
    "zooming": {
      "type": "x",
    },
  },
  "drilldown": {
    "activeDataLabelStyle": {
      "color": "#000000",
    },
    "breadcrumbs": {
      "buttonTheme": {
        "fill": "#ffcb05",
        "states": {
          "hover": {
            "fill": "#f2b900",
            "stroke": "#f2b900",
          },
        },
        "stroke": "#ffcb05",
        "style": {
          "color": "#3a4356",
        },
      },
    },
  },
  "legend": {
    "align": "center",
    "backgroundColor": "transparent",
    "borderColor": "transparent",
    "borderWidth": 0,
    "enabled": true,
    "itemStyle": {
      "color": "#000000",
      "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
      "fontSize": "13px",
      "fontWeight": "normal",
      "pointerEvents": "auto",
      "textOutline": "none",
    },
    "layout": "horizontal",
    "symbolRadius": 0,
    "verticalAlign": "bottom",
  },
  "plotOptions": {
    "boxplot": {
      "dataLabels": {
        "align": "center",
        "enabled": false,
        "formatter": [Function],
        "rotation": 0,
        "style": {
          "fontSize": "13px",
          "fontWeight": "normal",
        },
      },
    },
    "scatter": {
      "dataLabels": {
        "enabled": false,
      },
    },
    "series": {
      "animation": {
        "duration": 600,
      },
      "boostThreshold": 0,
      "dataLabels": {
        "align": "center",
        "enabled": false,
        "formatter": [Function],
        "rotation": 0,
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
        },
      },
      "grouping": false,
      "point": {
        "events": {
          "click": [Function],
        },
      },
      "softThreshold": false,
      "stickyTracking": false,
      "turboThreshold": 0,
    },
  },
  "series": [
    {
      "color": "#0090a1",
      "data": [
        {
          "blur": false,
          "color": "#0090a1",
          "custom": {
            "xValue": [
              "Apple Mac Desktops",
            ],
          },
          "fillColor": "#00cee6",
          "high": 1192.295066833496,
          "innerBoxColor": "#00afc4",
          "low": -446.05875396728516,
          "median": 335.3749694824219,
          "q1": 168.3239288330078,
          "q3": 577.9123840332031,
          "selected": false,
        },
        {
          "blur": false,
          "color": "#0090a1",
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "fillColor": "#00cee6",
          "high": 1167.9941291809082,
          "innerBoxColor": "#00afc4",
          "low": -455.2584037780761,
          "median": 281.7227478027344,
          "q1": 153.461296081543,
          "q3": 559.2744293212891,
          "selected": false,
        },
      ],
      "fillOpacity": 0.8,
      "legendIndex": 0,
      "maxPointWidth": 42,
      "medianWidth": 1,
      "minPointWidth": 20,
      "name": "Cost",
      "showInLegend": true,
      "strokeOpacity": 0.8,
      "title": "Cost",
      "whiskerLength": "100%",
      "whiskerWidth": 1,
      "yAxis": 0,
    },
    {
      "color": "#0090a1",
      "data": [
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Desktops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 0,
          "y": 1232.0960693359375,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Desktops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 0,
          "y": 1408.0069580078125,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 1,
          "y": 1181.6976318359375,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 1,
          "y": 1200,
        },
        {
          "blur": false,
          "custom": {
            "xValue": [
              "Apple Mac Laptops",
            ],
          },
          "marker": {
            "lineColor": "#0090a1",
          },
          "x": 1,
          "y": 1222.219970703125,
        },
      ],
      "legendIndex": 1,
      "marker": {
        "enabled": true,
        "fillColor": "transparent",
        "lineColor": "#0090a1",
        "lineWidth": 1,
      },
      "name": "Cost",
      "showInLegend": false,
      "title": "Cost",
      "type": "scatter",
    },
  ],
  "title": {
    "text": null,
  },
  "tooltip": {
    "animation": false,
    "backgroundColor": "#FFFFFF",
    "borderColor": "#CCCCCC",
    "borderRadius": 10,
    "borderWidth": 1,
    "formatter": [Function],
    "style": {
      "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
    },
    "useHTML": true,
  },
  "xAxis": [
    {
      "categories": [
        "Apple Mac Desktops",
        "Apple Mac Laptops",
      ],
      "endOnTick": true,
      "gridLineColor": "#d1d1d7",
      "gridLineDashStyle": "Dot",
      "gridLineWidth": 0,
      "labels": {
        "autoRotation": [
          -10,
          -20,
          -30,
          -40,
          -50,
          -60,
          -70,
          -80,
          -90,
        ],
        "enabled": true,
        "formatter": [Function],
        "overflow": "none",
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
      },
      "lineColor": "#d1d1d7",
      "max": null,
      "min": null,
      "minorGridLineColor": "#d1d1d7",
      "minorGridLineDashStyle": "Dot",
      "minorTickColor": "#d1d1d7",
      "minorTickLength": 0,
      "plotBands": undefined,
      "startOnTick": true,
      "tickColor": "#d1d1d7",
      "tickLength": 0,
      "tickmarkPlacement": "on",
      "title": {
        "margin": 25,
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
        },
        "text": null,
      },
    },
  ],
  "yAxis": [
    {
      "gridLineColor": "#d1d1d7",
      "gridLineDashStyle": "Dot",
      "gridLineWidth": 1,
      "labels": {
        "autoRotation": [
          -10,
          -20,
          -30,
          -40,
          -50,
          -60,
          -70,
          -80,
          -90,
        ],
        "enabled": true,
        "formatter": [Function],
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
      },
      "lineColor": "#d1d1d7",
      "lineWidth": 1,
      "max": 1192.295066833496,
      "min": -455.2584037780761,
      "minorGridLineColor": "#d1d1d7",
      "minorGridLineDashStyle": "Dot",
      "minorGridLineWidth": 0,
      "minorTickColor": "#d1d1d7",
      "plotBands": undefined,
      "tickColor": "#d1d1d7",
      "tickInterval": null,
      "tickmarkPlacement": "on",
      "title": {
        "margin": 25,
        "style": {
          "color": "#000000",
          "fontFamily": ""Open Sans","Roboto","Helvetica","Arial",sans-serif",
          "fontSize": "13px",
          "fontWeight": "normal",
          "pointerEvents": "none",
          "textOutline": "none",
        },
        "text": null,
      },
    },
  ],
}
`;
